webcrypto/subtle_crypto.go

```diff
@@ -597,11 +605,69 @@ func (sc *SubtleCrypto) DeriveKey(
 // using `SubtleCrypto.ImportKey`.
 //
 // The `length` parameter is the number of bits to derive. The number should be a multiple of 8.
-//
-//nolint:revive // remove the nolint directive when the method is implemented
 func (sc *SubtleCrypto) DeriveBits(algorithm goja.Value, baseKey goja.Value, length int) *goja.Promise {
-	// TODO: implementation
-	return nil
+	promise, resolve, reject := promises.New(sc.vu)
+	rt := sc.vu.Runtime()
+
+	var publicKey, privateKey CryptoKey
+	var algName string
+
+	err := func() error {
+		if err := rt.ExportTo(baseKey, &privateKey); err != nil {
+			return NewError(InvalidAccessError, "provided baseKey is not a valid CryptoKey")
+		}
+
+		if privateKey.Type != PrivateCryptoKeyType {
+			return NewError(InvalidAccessError, fmt.Sprintf("provided baseKey is not a private key: %v", privateKey))
+		}
+
+		alg := algorithm.ToObject(rt)
+
+		pcValue := alg.Get("public")
+		if err := rt.ExportTo(pcValue, &publicKey); err != nil {
+			return NewError(InvalidAccessError, "algorithm's public is not a valid CryptoKey")
+		}
+
+		algName = alg.Get("name").String()
+
+		if publicKey.Type != PublicCryptoKeyType {
+			return NewError(InvalidAccessError, "algorithm's public key is not a public key")
+		}
+
+		return nil
+	}()
+	if err != nil {
+		reject(err)
+		return promise
+	}
+
+	go func() {
+		result, err := func() ([]byte, error) {
+			deriver, err := newBitsDeriver(algName)
+			if err != nil {
+				return nil, err
+			}
+
+			b, err := deriver(privateKey, publicKey)
+			if err != nil {
+				return nil, NewError(OperationError, err.Error())
+			}
+
+			if len(b) < length/8 {
+				return nil, NewError(OperationError, "length is too large")
+			}
+
+			return b[:length/8], nil
+		}()
+		if err != nil {
+			reject(err)
+			return
+		}
+
+		resolve(rt.NewArrayBuffer(result))
```

Comment:
`rt.NewArrayBuffer` uses the runtime , but this is done in the event loop.

Given your code I would recommend moving away from `promises.New` to `goja.NewPromise` and calling `vu.RegisterCallback` just before the goroutine creation. Then calling from the if below in hte callback as in
```suggestion
callback(func() error {
		if err != nil {
			reject(err)
			return
		}

		resolve(rt.NewArrayBuffer(result))
})
```

This way this whole thing will be executed on the event loop and you can both use the runtime but also the not wrapped resolve/reject.

Reply:
Let me clarify :thinking: 

 Is the idea to go out of the vanilla go routine because we still use runtime? (`rt.NewArrayBuffer`)

Reply:
The clarification question still present, but the change is here https://github.com/grafana/xk6-webcrypto/pull/67/commits/a15ebeb484aa65e6c807461b8161b5fabca1aadd

Reply:
the problem is that `rt.NewArrayBuffer` is *not* happening on the event loop in the original code. It happens on a goroutine that isn't running on the event loop. At which point you have a race as this is accessing the runtime, again off the event loop. 

`promises.New` is a wrapper around `goja.NewPromises` and the event loop, so that the `resolve` and `reject` calls can be called on the event loop without the users specifically working with `RegisterCallback` and co.

In this case though as the argument to resolve needs to be an array buffer - you already need to run the `rt.NewArryaBuffer` part *on* the event loop - so you need to use `RegisterCallback` either way. 

Also in this case as *before* you make the goroutine you are on the event loop, the usefulness of `promises.New`'s `reject` being able to be called off the event loop isn't useful and neither is the calling of `resolve` if you need to *again* go back to the event loop to make the ArrayBuffer.

As such my proposal was to *not* use it at all and instead use RegisterCallback.

```pseudo-go
func something() promise {
  // here we are on the event loop
  p, res, rej := promises.New()
  if err != nil {
    rej(err); // this now actually just queues a job on the eventloop, and then when the callstack is empty will actuall reject the promise. 
   // not a deal breaker by any problem, but 
    return p;
  }

  p, res, rej := goja.NewPromise();

  if err != nil {
    rej(err); // this actually just rejects it outright, so the event loop doesn't need to go through this
    return p 
  }
  callback := vu.RegisterCallback()
  go func() {
    // do async work no actual usage of `rt`
   callback(func() error {
     if err !=nil {
       rej(err); // this already is on the event loop as it is the callback - so no need to queue a new job
       return
    }
    res(rt.newArrayBuffer()); // here we are on the event loop so we can use rt no problem.
  }) 
  }
  return p
}
```

In the above case if you drop `promises.New` - you will also never need to do `RegisterCallback` and directly interact with the event loop in the case you have an error.

tl;dr: You can still use `promises.New` - you just will make additional eventloop jobs that will not be useful and you in practice still need to `RegisterCallback` . If we didn't need to do work with the `runtime` before we can resolve/reject the promise - we could've used `promises.New`

Reply:
Okay, if I'm reading this right, the https://github.com/grafana/xk6-webcrypto/commit/a15ebeb484aa65e6c807461b8161b5fabca1aadd changes are fine, but we also need to refactor a bit other places where `rt` accesses inside the plain go routines, e.g.:

* https://github.com/grafana/xk6-webcrypto/blob/main/webcrypto/subtle_crypto.go#L193
* https://github.com/grafana/xk6-webcrypto/blob/main/webcrypto/subtle_crypto.go#L293
* etc

Reply:
Looking at the changes - you now do not have a gorotuine - you directly use the callback, so nohting happens of the event loop .

Otherwise - yes the those are other cases of the same problem that we haven't caught.

Reply:
How about this https://github.com/grafana/xk6-webcrypto/pull/67/commits/411de407ede2efd2e51547500d8d9c91cc291dcc ? reject is still out of the callback, but my understanding this should be fine :thinking:

Reply:
Okay, as a summary of the internal conversation with @mstoykov 

We should access the runtime (`rt`) only through the event loop (in this example, it's inside the `callback`), but if we directly use `rt.NewPromise` the both `resolve` and `reject` should be executed on top of the event loop, so the result version https://github.com/grafana/xk6-webcrypto/pull/67/files#diff-9e8a31e9776cf152256880a32e1c5751bf67322c23c0cb096d71cf0633d8fe14R665-R673

That way, we'll be protected from the possible races.